File: algo/valid-number/isNumber.c
typedef enum {
    TOKEN_BLANK = 0,
    TOKEN_SIGN, 
    TOKEN_NUM,
    TOKEN_INT,
    TOKEN_EPX
} TOKEN_KIND;

bool eat(char** s, TOKEN_KIND kind) {
    char* p = *s;
    bool hasNum;
    char* t;
    
    switch (kind) {
        case TOKEN_BLANK:
            while (*p == ' ') {
                ++p;
            }
            if (p == *s) {
                return false;
            }
            break;
        case TOKEN_SIGN:
            if (*p == '+' || *p == '-') {
                ++p;
            } else {
                return false;
            }
            break;            
        case TOKEN_NUM:
            hasNum = false;
            while (*p >= '0' && *p <= '9') {
                ++p;
            }
            if (p != *s) {
                hasNum = true;
            }
            if (*p == '.') {
                ++p;
            }
            t = p;
            while (*p >= '0' && *p <= '9') {
                ++p;
            }
            if (t == p && !hasNum) {
                return false;
            }
            break;
        case TOKEN_INT:
            while (*p >= '0' && *p <= '9') {
                ++p;
            }
            if (p == *s) {
                return false;
            }
            break;
        case TOKEN_EPX:
            if (*p == 'e') {
                ++p;
            } else {
                return false;
            }
            break;
        default:
            return false;
    }
    
    *s = p;
    return true;
}

bool isNumber(char* s) {
    if (s == NULL) {
        return false;
    }
    
    eat(&s, TOKEN_BLANK);
    eat(&s, TOKEN_SIGN);
    if (eat(&s, TOKEN_NUM)) {
        if (eat(&s, TOKEN_EPX)) {
            eat(&s, TOKEN_SIGN);
            if (!eat(&s, TOKEN_INT)) {
                return false;
            }
        }
        eat(&s, TOKEN_BLANK);
        if (*s == '\0') {
            return true;
        }
    }
    return false;
}

File: algo/valid-number/isNumber.go
type TOKEN_KIND int

const (
    TOKEN_BLANK TOKEN_KIND = iota
    TOKEN_SIGN
    TOKEN_NUM
    TOKEN_INT
    TOKEN_EPX
)

func eat(s *[]byte, kind TOKEN_KIND) bool {
    if len(*s) == 0 {
        return false
    }
    
    switch kind {
        case TOKEN_BLANK:
            for i, b := range *s {
                if b != ' ' {
                    *s = (*s)[i:]
                    if i == 0 {
                        return false
                    } else {
                        return true
                    }
                }
            }
            *s = []byte{}
            return true
        case TOKEN_SIGN:
            if (*s)[0] == '+' || (*s)[0] == '-' {
                *s = (*s)[1:]
                return true
            } else {
                return false
            }
        case TOKEN_NUM:
            ok1 := eat(s, TOKEN_INT)
            if len(*s) > 0 {
                if (*s)[0] == '.' {
                    *s = (*s)[1:]
                }
            }
            ok2 := eat(s, TOKEN_INT)
            if !ok1 && !ok2 {
                fmt.Println(1)
                return false
            } else {
                return true
            }
        case TOKEN_INT:
            for i, b := range *s {
                if b < '0' || b > '9' {
                    *s = (*s)[i:]
                    if i == 0 {
                        return false
                    } else {
                        return true
                    }
                }
            }
            *s = []byte{}
            return true
        case TOKEN_EPX:
            if (*s)[0] == 'e' {
                *s = (*s)[1:]
                return true
            } else {
                return false
            }
        default:
            return false
    }
}

func isNumber(s string) bool {
    bs := []byte(s)
    eat(&bs, TOKEN_BLANK)
    eat(&bs, TOKEN_SIGN)
    
    if eat(&bs, TOKEN_NUM) {
        if eat(&bs, TOKEN_EPX) {
            eat(&bs, TOKEN_SIGN)
            if !eat(&bs, TOKEN_INT) {
                return false
            }
        }
        eat(&bs, TOKEN_BLANK)
        if len(bs) == 0 {
            return true
        }
    }
    return false
}